packages/swarm-components/src/NumericalInput.jsx

```diff
@@ -2,47 +2,185 @@
+	state = {
+		value: this.props.value || this.props.value === 0 ? this.props.value : null,
+		isFieldFocused: false,
+	};
 
-    const { className = '', ...other } = props;
+    // internal value updater when passed a value via props
+	static getDerivedStateFromProps(nextProps: Props, prevState: State) {
+		const isNewValue = nextProps.onChange && nextProps.value !== prevState.value;
 
-    return (
-        <div data-swarm-number-input={getStatus(props)} className={className} >
-            <input
-                data-swarm-number-input-field
-                type="number"
-                pattern="[0-9]*"
-                {...other}
-            />
-            <button data-swarm-number-input-increment onClick={props.onIncrament} ><Icon shape="plus" size="xs" /></button>
-            <button data-swarm-number-input-decrement onClick={props.onDecrement} ><Icon shape="minus" size="xs" /></button>
-        </div>
-    );
-};
+		return {
+			value: isNewValue ? nextProps.value : prevState.value,
+		};
+	}
+
+	_updateValueByStep = (isIncreasing: boolean) => {
+		const { min, max, step } = this.props;
+		const currentVal = this.state.value || min || 0;
+
+		const newValue = isIncreasing ? currentVal + step : currentVal - step;
+		const minConstrained = Math.max(newValue, this.props.min || -Infinity);
+		return Math.min(minConstrained, max || Infinity);
+	};
+
+    // private value updater for keyboard and button interactions
+	_updateValue = (value: Value) => {
+		this.setState(() => ({ value }));
+
+		if (this.props.onChange) {
+			this.props.onChange(value);
+		}
+	};
+
+    // this signals blur for the inpurt and the two button controls
+    // only registers a blur when all elements are blurred
+	onBlur = (e: SyntheticInputEvent<HTMLInputElement>) => {
+		const formControls = [this.fauxInputEl, this.decrementBtnEl, this.incrementBtnEl];
+		if (formControls.every(c => c !== document.activeElement)) {
+			this.setState(() => ({ isFieldFocused: false }));
+            if (this.props.onBlur) {
+                this.props.onBlur(e);
+            }
+		}
+	};
+
+	onChange = (e: SyntheticInputEvent<HTMLInputElement>) => {
+		const newValue = e.target.value ? parseInt(e.target.value, 10) : null;
+		if (Number.isNaN(newValue)) {
+			return;
+		}
+		this._updateValue(newValue);
+	};
+
+    // this signals focus for the inpurt and the two button controls
+	onFocus = (e: SyntheticFocusEvent<HTMLInputElement>) => {
+		this.setState(() => ({ isFieldFocused: true }));
+	};
+
+	onKeyDown = (e: SyntheticKeyboardEvent<HTMLInputElement>) => {
+		// Disable the 'e' or 'E' values because we don't
+		// support scientific notation at the moment
+		if (e.key.toLowerCase() === 'e') {
+			e.preventDefault();
+		}
+	};
+
+    // the event for the increment button
+	incrementAction = (e: SyntheticInputEvent<HTMLInputElement>) => {
+		e.preventDefault();
+		this._updateValue(this._updateValueByStep(true));
```

Comment:
definitely could... but also 'one function one responsibility' principle supports this approach...

That set, `_updateValueByStep` is a little misleading, since it's actually _getting_ a value, not doing the update, so `_getNextStepValue` might be a better name to indicate it's a pure function